### PR TITLE
Step 37 modernization

### PR DESCRIPTION
**Pull request by drwells**
Part of #6774.

I am having some trouble getting the second non-homogeneous Dirichlet boundary condition implementation (i.e., the one with blocks) to work (multiplying the vector of inhomogeneities by the linear operator doesn't scale its values, so the RHS vector values are messed up), so this is WIP until I figure that out.

### CODE REVIEW

File: examples/step-37/doc/results.dox [hunk 1]
```diff
@@ -556,22 +559,25 @@ values as follows, assuming the Dirichlet values have been interpolated into
...
+  FEEvaluation<dim, degree_finite_element> phi(*system_matrix.get_matrix_free());
+  for (unsigned int cell = 0;
+       cell < system_matrix.get_matrix_free()->n_macro_cells();
+       ++cell)
     {
       phi.reinit(cell);
       phi.read_dof_values_plain(solution);
-      phi.evaluate(false, true);
-      for (unsigned int q=0; q<phi.n_q_points; ++q)
+      phi.evaluate(true, true);
```
Comment (davydden) [inline examples/step-37/doc/results.dox]: you don't need to change this to `true` here as you only get `phi.get_gradient()`, so it does not need to know how to get values.

**Comment by tjhei** [comment]
Would it make sense to put the different variants somewhere into tests/ when we get them to work? I also wouldn't mind referencing something like "see tests/matrix_free/step37-non-hom_01.cc" in the example.

**Comment by drwells** [comment]
This is a good idea. I was not able to find any tests that use blocks in this way so we need something like it regardless.

File: examples/step-37/doc/results.dox (hunk 1 above)
Comment (drwells) [reply examples/step-37/doc/results.dox]: You're right. I fixed it.

**Comment by kronbichler** [comment]
I do not have the time to look into it in detail right now, but I agree with @drwells that the two options should be equivalent. Do we know where exactly they differ, i.e., is any of the norms you output different? We need to find out whether the problem appears in the index handling where we apply constraints (`preprocess_constraints` and `postprocess_constraints`) or whether the wrong data is read within the `FEEvaluation::read_dof_values` calls.

A related problem with the current strategy, while easier to code, is the fact that we also assemble a contribution to the Dirichlet constrained degrees of freedom, whereas we in fact would like to have zero contributions on the right hand side for the Dirichlet constrained DoFs. But that should apply to both strategies, so it cannot be the cause for the problem. The correct way would be to use different `FEEvaluation` objects for `read_dof_values` and `distribute_local_to_global` in the inhomogeneous right hand side operator parts.

**Comment by drwells** [comment]
@davydden That version is also suggested by the tutorial and I confirmed that it works. I'll put up a PR that uses it in a test program (as suggested by @tjhei) shortly.

@kronbichler There is something off about the way that the block-based approach enforces boundary values. Here is a plot of the max error per cell for the block case:
errors-bad
and here it is for the completely separate second Laplace operator:
errors-good

Here is console output for the separate operator (good) case:
```
Cycle 4
Number of degrees of freedom: 2146689
solution.l2_norm() = 313.538
system_rhs.l2_norm() = 4.85369
system_rhs.l2_norm() = 4.85486
Total setup time               (wall) 37.0248s
Time solve (6 iterations)  (CPU/wall) 9.93996s/10.0512s
max error: 1.48143e-06
error ratio: 7.99298
```

where error ratio should be about 8 since we have quadratic elements. Here is the output for the combined operator (not working) case:

```
Cycle 4
Number of degrees of freedom: 2146689
solution.l2_norm() = 313.538
system_rhs.l2_norm() = 0
system_rhs.l2_norm() = 0.00796328
Total setup time               (wall) 34.9419s
Time solve (6 iterations)  (CPU/wall) 9.19891s/9.3158s
max error: 2.46833
error ratio: 0.9947
```
i.e., it looks like the vector we get from `dirichlet_constraints.distribute()` is in the null space of the Laplace operator, which doesn't make any sense. My best guess is that the second block is using the wrong affine constraints object at some point.

**Comment by davydden** [comment]
I checked the usage of `selected_rows` (what you provide, `{1}`) in `operators.h` and it looks reasonable, I did not spot any bugs there... We use it to setup `FEEvaluation`:
```
    FEEvaluation<dim, fe_degree, n_q_points_1d, n_components, Number> phi(
      data, this->selected_rows[0]);
```
and also for constraints like
```
    for (unsigned int j = 0; j < n_blocks(dst); ++j) // <---- will give 1 block for non-block vector
      {
        const std::vector<unsigned int> &constrained_dofs =
          data->get_constrained_dofs(selected_rows[j]);
        for (unsigned int i = 0; i < constrained_dofs.size(); ++i)
          subblock(dst, j).local_element(constrained_dofs[i]) +=
            subblock(src, j).local_element(constrained_dofs[i]);
      }
```

But of course your code under `if 1` looks perfectly valid.

**Comment by kronbichler** [comment]
I found the reason for the problem: In `step-37`, we do manually implement `LaplaceOperator::local_apply`, rather than going through the implemented version. An of course, we do not have the facilities to choose `this->selected_rows[0]` enabled there. I do not have an overly strong opinion as to whether include it or not, but I somewhat lean to not use that concept in the first tutorial of matrix-free computations. Thus, the second option outlined by @drwells (else case in the define) seems more appropriate. Opinions?

**Comment by drwells** [comment]
Thanks for looking into this: this explanation makes sense.

We would have to modify `LaplaceOperator` a lot to make it work with blocks, so I think it would be best to go with the simpler alternative in this tutorial.

**Comment by drwells** [comment]
I went ahead and updated the second example to no longer use blocks. It explains that we would need to change LaplaceOperator and refers the reader to `MatrixFreeOperators::Base` to read more.

File: examples/step-37/step-37.cc [hunk 2]
```diff
@@ -1132,13 +1106,22 @@ namespace Step37
...
+  // refinement process. In addition, we use a compression algorithm that is
+  // optimized for speed rather than disk usage. The default setting (which
+  // optimizes for disk usage) makes saving the output take about 4 times as
+  // long as running the linear solver, while setting
+  // DataOutBase::VtkFlags::compression_level to
+  // DataOutBase::VtkFlags::best_speed lowers this to only one fourth the time
+  // of the linear solve.
+  //
+  // We disable the output when the mesh gets too large. Note that a variant
+  // of program has been run on hundreds of thousands MPI ranks with as many
```
Comment (kronbichler) [inline examples/step-37/step-37.cc]: This is pre-existing, but would you rephrase "a variant of this program has been run..."

**Pull request merged**